compiler/rustc_monomorphize/src/collector/abi_check.rs

```diff
@@ -0,0 +1,169 @@
+            tcx.fn_abi_of_instance(param_env.and((instance, ty::List::empty())))
+        }
+        _ => {
+            panic!("Invalid function call");
+        }
+    };
+
+    let Ok(callee_abi) = callee_abi else {
+        // ABI failed to compute; this will not get through codegen.
+        return vec![];
+    };
+    do_check_abi(tcx, callee_abi, caller, |required_feature| {
+        answer.push(required_feature.to_string());
+    });
+    answer
+}
+
+/// Checks that a call expression does not try to pass a vector-passed argument which requires a
+/// target feature that the caller does not have, as doing so causes UB because of ABI mismatch.
+pub(super) fn check_call_site_abi<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    callee: Ty<'tcx>,
+    span: Span,
+    caller: InstanceKind<'tcx>,
+) {
+    if callee.fn_sig(tcx).abi().is_rust() {
+        // "Rust" ABI never passes arguments in vector registers.
+        return;
+    }
+    for required_feature in tcx.call_site_abi_missing_features((callee, caller.def_id())) {
+        tcx.emit_node_span_lint(
+            ABI_UNSUPPORTED_VECTOR_TYPES,
+            CRATE_HIR_ID,
+            span,
+            AbiErrorDisabledVectorTypeCall { span, required_feature },
+        );
+    }
+}
+
+pub(super) fn provide(providers: &mut Providers) {
+    *providers = Providers { check_instance_abi, call_site_abi_missing_features, ..*providers }
+}
+
+pub(super) fn should_check_instance_abi<'tcx>(tcx: TyCtxt<'tcx>, instance: Instance<'tcx>) -> bool {
+    // We do not need to check the instance for feature-dependent ABI if we can determine that the
+    // function has "Rust" ABI, which is known not to be feature-dependent.
+    // Note that the check is still correct on Rust ABI functions, but somewhat expensive. Hence,
+    // checking for "Rust" ABI is just an optimization.
+    // We also avoid to try to determine the type of the instance, as doing so involves running a
+    // query that does not usually run for unchanged functions in incremental builds.
+    match instance.def {
+        InstanceKind::Item(def)
+        | InstanceKind::ReifyShim(def, ..)
+        | InstanceKind::FnPtrShim(def, ..)
+        | InstanceKind::Virtual(def, ..)
+        | InstanceKind::VTableShim(def) => {
+            // fn_sig ICEs on defs that are not functions.
+            if matches!(tcx.def_kind(def), DefKind::Fn | DefKind::AssocFn) {
+                !tcx.fn_sig(def).skip_binder().abi().is_rust()
+            } else if matches!(tcx.def_kind(def), DefKind::Ctor(..)) {
+                // Struct constructors are certainly always fine.
+                false
+            } else {
+                true
+            }
+        }
+        InstanceKind::DropGlue(..)
+        | InstanceKind::CloneShim(..)
+        | InstanceKind::FnPtrAddrShim(..)
+        | InstanceKind::AsyncDropGlueCtorShim(..)
+        | InstanceKind::ThreadLocalShim(..) => false,
+        InstanceKind::Intrinsic(..)
```

Comment:
Intrinsics are not actual function calls and don't have an ABI in that sense, so I think they can also be skipped.

Reply:
Given closures also use rust-call, I'd assume so can the corresponding shims. Changed.